chore: refactor lambdas

joostvanderborg opened the pull request:
Refactor lambdas:
* convert functions to typescript
* use generated lambda function classes in apifunction
* Adapt ApiStack to use generated classes. Use shared ts files from a
single folder.
* remove unused packages / node_modules from lambdas
* Bundle mustache templates in regular lambda bundling

Simplify installation process for lambdas (no separate install step)
and update dependencies from the projen project.

Commit:
chore(deps): upgrade dependencies

Upgrades project dependencies. See details in [workflow run].

[Workflow Run]: https://github.com/GemeenteNijmegen/mijn-nijmegen/actions/runs/4684123751

------

*Automatically created by projen via the "upgrade-acceptance" workflow*

Signed-off-by: github-actions <github-actions@github.com>

github-actions[bot] committed:
Merge pull request #411 from GemeenteNijmegen/github-actions/upgrade-acceptance

chore(deps): upgrade dependencies

Commit:
chore(deps): upgrade dependencies

Upgrades project dependencies. See details in [workflow run].

[Workflow Run]: https://github.com/GemeenteNijmegen/mijn-nijmegen/actions/runs/4695006023

------

*Automatically created by projen via the "upgrade-acceptance" workflow*

Signed-off-by: github-actions <github-actions@github.com>

github-actions[bot] committed:
Merge pull request #414 from GemeenteNijmegen/github-actions/upgrade-acceptance

chore(deps): upgrade dependencies

Commit:
chore(deps): upgrade dependencies

Upgrades project dependencies. See details in [workflow run].

[Workflow Run]: https://github.com/GemeenteNijmegen/mijn-nijmegen/actions/runs/4704955180

------

*Automatically created by projen via the "upgrade-acceptance" workflow*

Signed-off-by: github-actions <github-actions@github.com>

github-actions[bot] committed:
Merge pull request #417 from GemeenteNijmegen/github-actions/upgrade-acceptance

chore(deps): upgrade dependencies

Commit:
chore(deps): upgrade dependencies

Upgrades project dependencies. See details in [workflow run].

[Workflow Run]: https://github.com/GemeenteNijmegen/mijn-nijmegen/actions/runs/4710498576

------

*Automatically created by projen via the "upgrade-acceptance" workflow*

Signed-off-by: github-actions <github-actions@github.com>

github-actions[bot] committed:
Merge pull request #418 from GemeenteNijmegen/github-actions/upgrade-acceptance

chore(deps): upgrade dependencies

Commit:
chore(deps): upgrade dependencies

Upgrades project dependencies. See details in [workflow run].

[Workflow Run]: https://github.com/GemeenteNijmegen/mijn-nijmegen/actions/runs/4716137117

------

*Automatically created by projen via the "upgrade-acceptance" workflow*

Signed-off-by: github-actions <github-actions@github.com>

github-actions[bot] committed:
Merge pull request #419 from GemeenteNijmegen/github-actions/upgrade-acceptance

chore(deps): upgrade dependencies

joostvanderborg committed:
feat: use typescript lambdas (#413)

* convert login function to typescript
* use generated lambda function classes in apifunction
* Adapt ApiStack to use generated classes. Use shared ts files from a
single folder.
* remove unused packages / node_modules from lambdas

Simplify installation process for lambdas (no separate install step)
and update dependencies from the projen project. Only the templates
are still shared and added in an install script.

joostvanderborg committed:
feat: Import and bundle mustache files (#421)

Treat mustache templates as regular imports, bundling them with the
rest of the code.

Commit:
chore(deps): upgrade dependencies

Upgrades project dependencies. See details in [workflow run].

[Workflow Run]: https://github.com/GemeenteNijmegen/mijn-nijmegen/actions/runs/4727145582

------

*Automatically created by projen via the "upgrade-acceptance" workflow*

Signed-off-by: github-actions <github-actions@github.com>

github-actions[bot] committed:
Merge pull request #423 from GemeenteNijmegen/github-actions/upgrade-acceptance

chore(deps): upgrade dependencies

joostvanderborg committed:
chore: bump (#424)

joostvanderborg committed:
fix: logout template includes weren't working (#425)

joostvanderborg committed:
Use env variable directly for endpoint (#426)

joostvanderborg committed:
fix: move setting endpoint to constructor (#427)

MartijnvDijk commented on src/ApiFunction.ts:
```diff
@@ -20,16 +22,17 @@ export interface ApiFunctionProps {
       handler: 'index.handler',
       description: props.description,
-      code: Lambda.Code.fromAsset(path.join(__dirname, props.codePath)),
+      code: Lambda.Code.fromInline('empty'), // Overwritten,
```
Hoe werkt dit?

joostvanderborg replied:
props.apiFunction wordt 'gevuld' met een verwijzing naar de class van de specifieke lambda (die auto-gegenereerd is door Projen, bijv. '[src/app/auth/auth-function.ts](https://github.com/GemeenteNijmegen/mijn-nijmegen/blob/852c9b1a279fa08dd07e2b395538c58d13cbcf3d/src/app/auth/auth-function.ts#L19)'. Dat is een variant van Lambda.Function. In de constructor wordt een instance van die class in this.lambda gemaakt, met de properties zoals hier te zien. Maar de auto-gegenereerde class bevat iets als dit:
```
     description: 'src/app/auth/auth.lambda.ts',
      ...props,
      runtime: new lambda.Runtime('nodejs18.x', lambda.RuntimeFamily.NODEJS),
      handler: 'index.handler',
      code: lambda.Code.fromAsset(path.join(__dirname, '../../../assets/app/auth/auth.lambda')),
```
Hier wordt eerst een description gezet, dan alle properties (zoals hier vanaf regel 31) toegevoegd, en daarná de runtime, handler, en code properties alsnog geset. In dit geval worden die dus overschreven, en maakt de waarde die hier op regel 35 staat niet uit.

MartijnvDijk commented on src/app/auth/handleRequest.ts:
```diff
@@ -0,0 +1,67 @@
+    const claims = await props.OpenIdConnect.authorize(props.queryStringParamCode, state, props.queryStringParamState);
+    if (claims) {
+      const bsn = new Bsn(claims.sub);
+      if (claims.hasOwnProperty('acr')) {
```
Waar komt 'acr' vandaan?

joostvanderborg replied:
Da's de claim die signicat teruggeeft met het betrouwbaarheidsniveau. Officieel de ['Authentication Context Class Reference'](https://openid.net/specs/openid-connect-core-1_0.html#acrSemantics).

marnixdessing commented on src/ApiFunction.ts:
```diff
@@ -7,7 +6,10 @@ import { Construct } from 'constructs';
+type T = Lambda.Function;
+
 export interface ApiFunctionProps {
+  apiFunction: {new(scope: Construct, id:string, props?: Lambda.FunctionProps): T };
```
Toch geen betere oplossing kunnen vinden? haha

marnixdessing replied:
Het zit nu in de properties ipv in de constructor dat is misschien wel iets netter inderdaad.

joostvanderborg replied:
Nee, wel wat mee gespeeld. Dit scheelt weer iets ;).

MartijnvDijk reviewed (commented):
Over mustache file kan ik weinig zeggen. Verder nog twee vragen.

marnixdessing commented on src/app/home/shared/footer.mustache:
Is deze file anders dan de ander header en footer?
Kunnen deze niet over alle lambdas heen gedeeld worden?

marnixdessing replied:
Ik zie dat ze wel geïmporteerd worden vanuit de directory waar `render.ts` ook leeft. Dan kunnen de andere waarschijnlijk weg toch?

joostvanderborg replied:
Die kunnen inderdaad weg verwacht ik. Ik heb niet opgelet met het builden van die shared folder / aanpassen van .projenrc, waardoor ze nu meegecommit zijn (Normaal werden ze in de build pas geplaatst in die folders). Zal ik nog even doen.

joostvanderborg committed:
chore: Remove unused shared files (#428)

Mustache shared files are now bundled and referenced from the
'render.ts' function. These are no longer necessary.

marnixdessing reviewed (approved):
Martijn en ik hebben een review gedaan. Extra uitleg toegevoegd en wat opgeruimd 👍
